js/modules/loki_file_server_api.js

```diff
@@ -40,6 +43,184 @@ class LokiFileServerAPI {
     return this._setOurDeviceMapping(authorisations, isPrimary);
   }
 
+  async getDeviceMappingForUsers(pubKeys) {
+    const users = await this._server.getUsersAnnotations(pubKeys);
+    return users;
+  }
+
+  async verifyUserObjectDeviceMap(
+    pubKeys,
+    isRequest,
+    iterator,
+    notFoundHandler
+  ) {
+    const users = await this.getDeviceMappingForUsers(pubKeys);
+    // log.info('verifyUserObjectDeviceMap Found', users.length, 'users')
+
+    // go through each user and find deviceMap annotations
+    users.forEach(user => {
+      let found = false;
+      if (user.annotations) {
+        user.annotations.forEach(note => {
+          if (note.type === 'network.loki.messenger.devicemapping') {
+            // is desired type
+            if (
+              (isRequest && note.value.isPrimary === '0') ||
+              (!isRequest && note.value.isPrimary !== '0')
+            ) {
+              const { authorisations } = note.value;
+              if (Array.isArray(authorisations)) {
+                authorisations.forEach(auth => {
+                  // log.info('devmap auth', auth);
+                  // only skip, if in secondary search mode
+                  if (
+                    isRequest &&
+                    auth.secondaryDevicePubKey !== user.username
+                  ) {
+                    // this is not the authorization we're looking for
+                    log.info(
+                      `Request and ${auth.secondaryDevicePubKey} != ${
+                        user.username
+                      }`
+                    );
+                    return;
+                  }
+                  // log.info('auth', auth);
+                  try {
+                    // request (secondary wants to be paired with this primary)
+                    // grant (primary approves this secondary)
+                    window.libloki.crypto.verifyPairingSignature(
+                      auth.primaryDevicePubKey,
+                      auth.secondaryDevicePubKey,
+                      dcodeIO.ByteBuffer.wrap(
+                        isRequest ? auth.requestSignature : auth.grantSignature,
+                        'base64'
+                      ).toArrayBuffer(),
+                      isRequest
```

Comment:
At this stage there should not be 'authorisation requests' in the mapping held by the server, so it's safe to say that all signatures can be verified with the GRANT type.
The request type signatures are only meant to be shared directly from a secondary device to a primary device.

Reply:
yes, depending on the isRequest flag, we're either checking the secondary to primary OR primary to secondary mapping. The isRequest flag denotes which way we're checking. We check both sides to make sure both are valid and in agreement. This was working code, so the server held this information at one point.

Reply:
Right, never mind what I said :P